webhooks: Add helper to auto-resolve topics.

Pritesh-30 opened the pull request:
This PR adds a common helper, `maybe_auto_resolve_topic`, that allows webhook integrations to auto-resolve a topic using the message id returned from `check_send_webhook_message`. The message id  is then used to update the message 
and mark the topic as resolved.
Used the helper in github to auto resolve when PR is closed or merged.
Fixes #34075.

<!-- Describe your pull request here.-->

**How changes were tested:**
I have added  test cases in `zerver/tests/test_webhooks_common.py` to check 
  - successful topic resolution,
  - no message sent case.
  - empty topic (general chat) is not resolved.
 
All tests pass locally.

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
Before
<img width="1137" height="221" alt="image" src="https://github.com/user-attachments/assets/2fa6634f-29a6-49a4-81ce-379b53fc364f" />
After
<img width="1139" height="310" alt="image" src="https://github.com/user-attachments/assets/27af6f3f-b053-4cd2-b1d6-0c93923cda4e" />
<img width="1131" height="248" alt="image" src="https://github.com/user-attachments/assets/e5643426-7307-460e-a48d-358f9c476ef8" />



<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

Pritesh-30 committed:
webhooks: Add helper to auto-resolve topics.

Add a common helper maybe_auto_resolve_topic that allows webhook
integrations to auto-resolve topics using the message ID returned.
Added tests covering resolved, unresolved and empty topic.

Fixes part of #34075.

Pritesh-30 commented:
@Niloth-p Can you review this PR?

Niloth-p commented:
Thanks for the PR, @Pritesh-30!

- It would be helpful to test this by using this in an integration. That can be added as a separate commit - either just for testing (if there are any blockers / limitations, or you just want to keep things simple for the moment), or ready enough for merging.
And do attach the before/after screenshots in the PR description.
- Thanks for accounting for the "general chat" topic. We generally shouldn't be resolving any user-specified custom topics. So, I'm not sure if we need to check for this in the function. But, yes, it's a good safeguard, to prevent wrong calls. Maybe we could make the safeguard more generic.
- Why would we call this function if `message_id` was None?
- When would we need to pass `should_resolve`?
- Also, what's the plan for how we're going to handle any future messages to that topic? They would get sent to the older topic.

Niloth-p commented:
Since the two PRs #37186 and #37183 are similar, it would be nice if you could both review each other's PRs, @DhruvShetty22. I'm tagging the latter under "buddy review", and will tag this one the same once the second commit is added. Cheers!

Pritesh-30 commented:
> * When would we need to pass `should_resolve`?
> * Also, what's the plan for how we're going to handle any future messages to that topic? They would get sent to the older topic.

- We should pass `should_resolve` for events like

    -   PR closed / PR merged
    -  Issue closed
- Yes thats right I have to think about it. Once a topic is resolved , any event message will be made in a new topic which we don't want. Currently I am working on integrating it with github. Currently I am facing a issue , when the PR is closed the topic is getting resolved but at the same time , another topic with unresolved status is getting created and moved
<img width="1271" height="543" alt="image" src="https://github.com/user-attachments/assets/12b32e57-1cfe-411e-8e65-1b2597928730" />. And any event message after that is going into unresolved topic .I am trying to find the cause.

Niloth-p commented:
> * We should pass `should_resolve` for events like
>   
>   * PR closed / PR merged
>   * Issue closed

Why would we be calling this function for any other events?

Pritesh-30 committed:
webhooks/github: Auto-resolve topics for closed & merged pull requests.

When github pull request is closed or merged, automatically mark
the corresponding stream as resolved.

Fixes part of #34075.

DhruvShetty22 commented on zerver/lib/webhooks/common.py:
```diff
@@ -210,6 +213,34 @@ def check_send_webhook_message(
+
+    topic = message.topic_name()
+    if not topic or topic.startswith(RESOLVED_TOPIC_PREFIX):
+        return
```
```suggestion
    if message_id is None or not topic_name or topic_name.startswith(RESOLVED_TOPIC_PREFIX):
        return
```

Pritesh-30 committed:
integrations/github: Add fixtures for issue closed event.

Issue closed event was already handled by generic event handlers
for issues.This commit adds coverage for it.

Fixes part of #34075.

Pritesh-30 committed:
webhooks/github: Auto-resolve topics for closed issues.

When github issue is closed automatically mark the
corresponding stream as resolved.

Fixes #34075.

Pritesh-30 commented:
@zulipbot remove "buddy review"

Pritesh-30 commented:
@zulipbot add "maintainer review"

Pritesh-30 commented:
@zulipbot remove "maintainer review" and @zulipbot add "mentor review"

Pritesh-30 commented:
@mounilKshah and @kennethnrk tagging you for review

kennethnrk commented:
LGTM! @zulipbot remove "mentor review"

laurynmm commented:
@Niloth-p would you be up for doing maintainer review here?

laurynmm commented:
@apoorvapendse would you be up for doing maintainer review on this one?

apoorvapendse reviewed (commented):
@Pritesh-30 is this still on your radar?

I notice that [#integrations > webhooks ignore resolved topics for follow-up events](https://chat.zulip.org/#narrow/channel/127-integrations/topic/webhooks.20ignore.20resolved.20topics.20for.20follow-up.20events/with/2468197) was the most recent discussion that happened around this.

Pritesh-30 commented:
Yes, this is still on my radar but hadn’t had a chance to look at it for a while. In the recent discussion, we confirmed that we need to send the message to the same resolved topic. We haven’t yet decided how exactly to implement this. Option A is to derive the topic name at send time use ✔ <topic> when the entity is closed/merged and <topic> otherwise, and pass that to `check_send_webhook_message`. However, it could override a human who manually unresolved the topic, since the bot would keep forcing it back to resolved on subsequent comments. Option B is to look up which variant already exists  query for ✔ <topic> and <topic> in the channel, send to whichever one is present, and only move the topic on an actual state transition. The downside is that this would add one query per message.
